src/zkAppWorker.ts

```diff
@@ -0,0 +1,162 @@
+import {
+  CircuitString,
+  type Field,
+  MerkleTree,
+  Poseidon,
+  type Proof,
+} from "snarkyjs";
+import { StateTracker } from "./program";
+import {
+  type TransitionRes,
+  type WorkerState,
+  transitionStateArgsSchema,
+} from "./types";
+import { INITIAL_STATE, MERKLE_TREE_HEIGHT, MerkleWitness20 } from "./utils";
+
+const state: WorkerState = {
+  tree: new MerkleTree(MERKLE_TREE_HEIGHT),
+  transitionIndex: 0n, // should increment on state transition and is used to index the tree
+  updateQueue: [],
+  executingUpdate: false,
+};
+
+setInterval(() => {
+  if (
+    !state.latestProof ||
+    state.executingUpdate ||
+    !state.updateQueue.length
+  ) {
+    return;
+  }
+
+  const nextUpdate = state.updateQueue.shift();
+  if (!nextUpdate) return;
+
+  state.executingUpdate = true;
+
+  nextUpdate(state.latestProof)
+    .then((proof) => {
+      state.latestProof = proof;
+
+      const message: ZkappWorkerReponse = {
+        resType: "proof-update",
+        id: 0,
+        data: proof.toJSON(),
+      };
+      postMessage(message);
+
+      state.executingUpdate = false;
+    })
+    .catch((err) => {
```

Comment:
Don't we need to "unshift" the last update in the queue if we catch an error?
And maybe implement some sort of re-try?

Reply:
I agree, but this is for a future PR, since this one is not concerned with implementation